### DISCUSSION TITLE
Issue with GPU Memory Allocation in Milvus!

### DISCUSSION THREAD

**Discussion by sail-amir** [milvus-io/milvus#36563] [Q&A and General discussion]
I have assigned 8 GPUs to Milvus, but when building the index, I noticed that only 4 out of the 8 GPUs show changes in memory usage. Despite having enough memory available across all GPUs, I encounter memory errors during the actual, likely due to this uneven distribution.

Do you have any idea why this is happening and how I can ensure that all 8 GPUs are fully utilized to avoid these memory errors?

**Comment by yhmo** [comment]
Milvus manages data by segments, a segment is a block of data with independent index. When you call collection.load() to load the data, the segments are dispatched to different GPU devices with load balance.
If your collection has only 4 segments, 4 GPUs will be used for the query.

**Reply by sail-amir** [reply]
thanks!
I’m encountering an "Out of Memory" error during search operation in Milvus, even though I have 8 GPUs, each with 80GB of memory. Upon checking the GPU usage, it appears that Milvus is fully occupying only 2 out of the 8 GPUs, while the other 6 GPUs remain largely unused. I already increased the max initpoolsize to 70gb and maxpoolsize to 80gb to allow it use everything.

Could you help me understand why Milvus is not utilizing all available GPUs, and how I can ensure that it distributes the workload evenly to avoid this memory issue?

---------------------------------------------------------------------------
MilvusException                           Traceback (most recent call last)
MilvusException: <MilvusException: (code=65535, message=fail to search on QueryNode 32: worker(32) query failed:  => failed to search: raft inner error: std::bad_alloc: out_of_memory: RMM failure at:/workspace/source/cmake_build/3rdparty_download/rmm-src/include/rmm/mr/device/pool_memory_resource.hpp:424: Maximum pool size exceeded at /workspace/source/internal/core/src/index/VectorMemIndex.cpp:625
)>

here is the image of gpu usage of milvus.
![image](https://github.com/user-attachments/assets/eaf05141-7484-416d-bdae-7195b64b8aed)

**Reply by yhmo** [reply]
Could you tell us how you deploy the milvus, especially the GPU configuration by following this doc: https://milvus.io/docs/install_standalone-docker-compose-gpu.md

I will call @Presburger to help.

**Comment by presburger** [comment]
@sail-amir Hello, welcome to using Milvus GPU. In the latest version, we modified the memory pool usage strategy. The memory pool is only used when query vectors enter Milvus, and the allocation of segments holding the base vectors does not use the memory pool. The main issue here is that the memory pool implementation does not shrink. Using the memory pool to allocate segment loading increases the likelihood of crashes. Therefore, keeping the `initPoolSize` and `maxPoolSize` at 2048 and 4096 is sufficient, unless the query load is extremely high, or the vector dimensions are very large, in which case you may consider increasing these two settings. As the memory pool grows larger, it will reduce the memory available for vector capacity, leaving less memory for segment loading.

**Comment by presburger** [comment]
@sail-amir The Milvus GPU version has a non-strict multi-GPU load balancing mechanism, but each GPU needs to be exclusively used by Milvus. If you run other tasks on some of the GPUs, it will affect Milvus's priority when allocating data to those GPUs.
